### PR TITLE
Add radial distance checks

### PR DESCRIPTION
**Pull request by andkerr**
This PR completes work started in #825, based on changes suggested in [this](https://github.com/glotzerlab/freud/pull/825#issuecomment-913825792) comment.

## Description
I add checks for appropriate radial distance arguments in the `RDF` and `NeighborList` class constructors, and unit tests that verify that invalid parameters raise appropriate error messages, rather than produce undefined behavior.

## Motivation and Context
Resolves Issue #821 and PR #825.

Will this PR require any changes to docs for the corresponding modules?

## How Has This Been Tested?
Existing code compiles. I am sorting out the unit tests (some of my checks for error messages do not pass right now).

## Screenshots
N/A

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds or improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation improvement (updates to user guides, docstrings, or developer docs)

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/glotzerlab/freud/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [ ] I have updated the documentation (if relevant).
- [x] I have added tests that cover my changes (if relevant).
- [x] All new and existing tests passed.
- [ ] I have updated the [credits](https://github.com/glotzerlab/freud/blob/master/doc/source/reference/credits.rst).
- [x] I have updated the [Changelog](https://github.com/glotzerlab/freud/blob/master/ChangeLog.md).

### CODE REVIEW

File: cpp/density/LocalDensity.cc [hunk 1]
```diff
@@ -12,7 +12,17 @@ namespace freud { namespace density {
 
 LocalDensity::LocalDensity(float r_max, float diameter)
     : m_box(box::Box()), m_r_max(r_max), m_diameter(diameter)
-{}
+{
+    if (r_max <= 0)
+    {
+        throw std::invalid_argument("Local density calculations require that r_max must be positive");
```
Comment (bdice) [inline cpp/density/LocalDensity.cc]: This suggestion matches the style of the existing error messages more closely. See #825 for other examples.
```suggestion
        throw std::invalid_argument("LocalDensity requires r_max to be positive.");
```

File: cpp/density/LocalDensity.cc [hunk 2]
```diff
@@ -12,7 +12,17 @@ namespace freud { namespace density {
...
+    }
+
+    if (diameter < 0)
+    {
+        throw std::invalid_argument("Local density calculations require that diameter must be non-negative");
```
Comment (bdice) [inline cpp/density/LocalDensity.cc]:
```suggestion
        throw std::invalid_argument("LocalDensity requires diameter to be non-negative.");
```

File: cpp/locality/NeighborQuery.h [hunk 3]
```diff
@@ -265,7 +265,16 @@ class NeighborQueryPerPointIterator : public NeighborPerPointIterator
         : NeighborPerPointIterator(query_point_idx), m_neighbor_query(neighbor_query),
           m_query_point(query_point), m_finished(false), m_r_max(r_max), m_r_min(r_min),
           m_exclude_ii(exclude_ii)
-    {}
+    {
+        if (r_max <= 0)
+        {
+            throw std::invalid_argument("Neighbor queries require that r_max must be positive.");
```
Comment (bdice) [inline cpp/locality/NeighborQuery.h]:
```suggestion
            throw std::invalid_argument("NeighborQuery requires r_max to be positive.");
```

File: cpp/locality/NeighborQuery.h [hunk 4]
```diff
@@ -265,7 +265,16 @@ class NeighborQueryPerPointIterator : public NeighborPerPointIterator
...
+        }
+        if (r_max <= r_min)
+        {
+            throw std::invalid_argument("Neighbor queries require that r_max must be greater than r_min.");
```
Comment (bdice) [inline cpp/locality/NeighborQuery.h]:
```suggestion
            throw std::invalid_argument("NeighborQuery requires that r_max must be greater than r_min.");
```

File: tests/test_density_LocalDensity.py [hunk 5]
```diff
@@ -94,3 +94,10 @@ def test_points_ne_query_points(self):
         ) / v_around
         correct_density = [cd0, cd1, 0]
         npt.assert_allclose(ld.density, correct_density, rtol=1e-4)
+
+    def test_invalid_radial_distances(self):
+        # make sure that invalid r_max and diameter arguments raise errrors
+        with pytest.raises(ValueError):
+            freud.density.LocalDensity(-1, 10)
+        with pytest.raises(ValueError):
+            freud.density.LocalDensity(2, 0)
```
Comment (bdice) [inline tests/test_density_LocalDensity.py]: Diameter is allowed to be zero. A negative value should raise an error.
```suggestion
            freud.density.LocalDensity(2, -1)
```

**Review by bdice — COMMENTED** [review]
Thanks for working on this!

**Comment by vyasr** [comment]
Thanks @andkerr!

**Comment by andkerr** [comment]
@bdice, I'm running into a problem that I haven't seen before. It seems like extra validation I added to the `LocalDensity` and `NeighborPointQueryIterator` constructors are causing pytest to abort on a fatal error when it tries to run the test I wrote to check that the appropriate errors are raised for invalid parameters. It looks like the problem is specifically with the additional checks in `LocalDensity.cc`. When I comment them out, the test I added simply fails with no error. 

'm not sure where my mistake is, I've done my best to match the form of other calls to `std::invalid_argument`. This is the error message I get:

Screen Shot 2021-11-21 at 2 52 05 PM

**Comment by bdice** [comment]
@andkerr It's because the constructor for `LocalDensity` is missing an `except +` in the Cython bindings: https://github.com/glotzerlab/freud/blob/94e8f2404c3b8f4976634e2ce42c7eaf474dcf88/freud/_density.pxd#L38

Compare to the RDF constructor, which has `except +`:
https://github.com/glotzerlab/freud/blob/94e8f2404c3b8f4976634e2ce42c7eaf474dcf88/freud/_density.pxd#L52

Please add `except +` to all the functions (probably all constructors and `compute` methods) that can raise an error. 👍

Cython docs: https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#error-return-values

**Comment by andkerr** [comment]
Great, I added the `except +` flag, tests are passing now.

File: cpp/locality/NeighborQuery.h [hunk 6]
```diff
@@ -265,7 +265,16 @@ class NeighborQueryPerPointIterator : public NeighborPerPointIterator
...
+    {
+        if (r_max <= 0)
+        {
+            throw std::invalid_argument("NeighborQuery requires r_max to be positive.");
+        }
+        if (r_max <= r_min)
+        {
+            throw std::invalid_argument("NeighborQuery requires that r_max must be greater than r_min.");
+        }
+    }
```
Comment (bdice) [inline cpp/locality/NeighborQuery.h:277]: Can we add Python tests for this? It probably needs to be tested using concrete subclasses like AABBQuery and LinkCell. We need to be sure we get `except +` on all the methods that can construct a NeighborQuery -- that might take some careful thought.

Comment (andkerr) [reply cpp/locality/NeighborQuery.h:277]: Sure thing, I'll take care of this.

**Comment by andkerr** [comment]
I added checks for appropriate errors when invalid radius arguments are passed to `AABB Query` and `LinkCell`. Both classes already had the necessary `except +` flags in the Cython declarations. I had to pass some additional argument to the calls to `query()` (`num_neighbors`, `mode`) to prevent Freud from throwing a RuntimeError, and am not sure if this was the correct choice.

File: cpp/density/LocalDensity.cc [hunk 7]
```diff
@@ -12,7 +12,17 @@ namespace freud { namespace density {
...
-{}
+{
+    if (r_max <= 0)
+    {
+        throw std::invalid_argument("LocalDensity requires r_max to be positive.");
+    }
+
+    if (diameter < 0)
+    {
+        throw std::invalid_argument("LocalDensity requires diameter to be non-negative");
```
Comment (bdice) [inline cpp/density/LocalDensity.cc]:
```suggestion
        throw std::invalid_argument("LocalDensity requires diameter to be non-negative.");
```

File: cpp/locality/NeighborQuery.h (hunk 6 above)
Comment (bdice) [reply cpp/locality/NeighborQuery.h:277]: @andkerr I changed the new tests a bit in ad4e6f4e9148796d7690fd749d85e812c2f2513d. It looks like you defined the tests in both AABBQuery and LinkCell classes, but the logic can be defined once in their shared parent class.

**Review by bdice — APPROVED** [review]
LGTM, I pushed some small tweaks. Automerging when tests pass.

File: tests/test_density_LocalDensity.py [hunk 8]
```diff
@@ -94,3 +94,10 @@ def test_points_ne_query_points(self):
         ) / v_around
         correct_density = [cd0, cd1, 0]
         npt.assert_allclose(ld.density, correct_density, rtol=1e-4)
+
+    def test_invalid_radial_distances(self):
+        # make sure that invalid r_max and diameter arguments raise errrors
```
Comment (bdice) [inline tests/test_density_LocalDensity.py]:
```suggestion
        """Ensure that invalid r_max and diameter arguments raise errors."""
```

**Comment by andkerr** [comment]
Nice!

**Pull request merged**